great_expectations/validator/metrics_calculator.py

```diff
@@ -97,10 +129,14 @@ def head(
         )
 
         if isinstance(
-            self._execution_engine, (PandasExecutionEngine, SqlAlchemyExecutionEngine)
+            self._execution_engine,
```

Comment:
this logic of "if this type then, else if this type ..." shouldn't live here, the dataframe should be the result of a shared method on the engines themselves. This kind of logic is why it is so hard to add new engines

Reply:
@donaldheppner This is different here.  The `MetricsCalculator.head()` is a convenience method that is supposed to return a `Pandas` `DataFrame`.  For `SqlAlchemy` and `Pandas`, the output of the metric directly plugs in to `pd.DataFrame()`, but for `Spark`, an additional data conditioning is needed.  Thanks.

Reply:
That's not the point. The point is that if we add a 4th engine, we need to modify this condition to account for it. It forces engine-specific logic into what should be an engine-agnostic class. Please make the change

Reply:
@donaldheppner Not necessarily.  The 4th engine might have the same way of conveniently supporting `pd.DataFrame`.  This logic has been in `MetricsCalculator` outside of this PR.  The only way to change this behavior and make it engine neutral is to do exactly the same as I did for `Query` metrics, and that is make the metric calculation output the same for all execution engines; then we will not have the Engine dependency in the called (i.e., `MetricsCalculator`).  Strictly speaking, this belongs in a separate (very small pull request), but I am happy to make it part of this one.  Please advise which way is best -- it is a rather small change either way.  Thank you!

Reply:
@donaldheppner I added the code to handle circular dependencies; the ExecutionEngine code was there before. Hence, the logical place for this is in a separate pull request (which can come before this one, by the way along with other V3 enhancements needed to support both V3 and ZEP -- if we wish to handle it this way; please advise). Thanks.